source/common/common/token_bucket.cc

```diff
@@ -0,0 +1,31 @@
+#include "common/common/token_bucket.h"
+
+#include <algorithm>
+
+namespace Envoy {
+
+TokenBucket::TokenBucket(uint64_t max_tokens, double tokens_per_sec)
+    : max_tokens_(max_tokens), refill_rate_(tokens_per_sec * nano), tokens_(max_tokens),
+      last_fill_(std::chrono::high_resolution_clock::now()) {}
+
+bool TokenBucket::consume(uint64_t tokens) {
+  const auto time_now = std::chrono::high_resolution_clock::now();
```

Comment:
You might want the clock to be a constructor parameter to allow dependency injection in testing to mock out time, see https://github.com/envoyproxy/envoy/blob/0efa18c36d6f789562690b5053c2c4b00987979e/test/mocks/common.h#L32.

Reply:
Just stumbled upon this PR. Not to show off, but here is my own implementation which does the things you might be looking for (based on a @black-adder's implementation in Go): https://github.com/jaegertracing/cpp-client/blob/master/src/jaegertracing/utils/RateLimiter.h.

Reply:
Hi @isaachier, Thank you for the example. I thought about this algorithm to be thread-safe as well, but it's not necessary for what we need at the moment. I might just create an interface to make it a bit more generic though.

Reply:
Actually ignore the mutex. The main point here is the use of a template for the clock type.

Reply:
Good point, but I'm afraid that it will make dependency injection somehow complicate. I recently changed it to use Envoy's MonotonicTimeSource interface.

https://github.com/envoyproxy/envoy/blob/master/include/envoy/common/time.h.